ballista/rust/scheduler/src/flight_sql.rs

```diff
@@ -0,0 +1,402 @@
+                    Status::internal(msg)
+                })?;
+            let status: JobStatus = match status {
+                Some(status) => status,
+                None => {
+                    let msg = format!("Error getting status for job {}!", job_id);
+                    error!("{}", msg);
+                    Err(Status::internal(msg))?
+                }
+            };
+            let status: job_status::Status = match status.status {
+                Some(status) => status,
+                None => {
+                    let msg = format!("Error getting status for job {}!", job_id);
+                    error!("{}", msg);
+                    Err(Status::internal(msg))?
+                }
+            };
+            let completed = match status {
+                job_status::Status::Queued(_) => continue,
+                job_status::Status::Running(_) => continue,
+                job_status::Status::Failed(e) => {
+                    Err(Status::internal(format!("Error building plan: {}", e.error)))?
+                }
+                job_status::Status::Completed(comp) => comp
+            };
+            let mut fieps: Vec<_> = vec![];
+            for loc in completed.partition_location.iter() {
+                let fetch = if let Some(ref id) = loc.partition_id {
+                    let fetch = protobuf::FetchPartition {
+                        job_id: id.job_id.clone(),
+                        stage_id: id.stage_id,
+                        partition_id: id.partition_id,
+                        path: loc.path.clone()
+                    };
+                    protobuf::Action {
+                        action_type: Some(protobuf::action::ActionType::FetchPartition(fetch)),
+                        settings: vec![],
+                    }
+                } else {
+                    Err(Status::internal(format!("Error getting partition it")))?
+                };
+                let authority = if let Some(ref md) = loc.executor_meta {
+                    // pub id: ::prost::alloc::string::String,
+                    format!("{}:{}", md.host, md.port)
+                } else {
+                    Err(Status::internal(format!("Error getting location")))?
+                };
+                if let Some(ref stats) = loc.partition_stats {
+                    num_rows += stats.num_rows;
+                    num_bytes += stats.num_bytes;
+                    // pub num_batches: i64,
+                } else {
+                    Err(Status::internal(format!("Error getting stats")))?
+                }
+                let loc = Location { uri: format!("grpc+tcp://{}", authority) };
+                let buf = fetch.encode_to_vec();
+                let ticket = Ticket { ticket: buf };
+                let fiep = FlightEndpoint {
+                    ticket: Some(ticket),
+                    location: vec![loc],
+                };
+                fieps.push(fiep);
+            }
+            break fieps;
```

Comment:
There seems no need to use a label here for the break?

Reply:
It's pretty neat. You can actually return a value from a `loop`, and that's what this is doing. I wasn't aware of it either, until this PR. https://doc.rust-lang.org/rust-by-example/flow_control/loop/return.html